exotations/task_maps/task_map/src/Point2Point.cpp

```diff
@@ -0,0 +1,232 @@
+    double cp, sp; // cos/sin(roll)
+    double cy, sy; // cos/sin(roll)
+
+    cr = std::cos(roll);
+    sr = std::sin(roll);
+
+    cp = std::cos(pitch);
+    sp = std::sin(pitch);
+
+    cy = std::cos(yaw);
+    sy = std::sin(yaw);
+
+    // Row 1
+    R(0,0) = cr*cp;
+    R(0,1) = cr*sp*sy - sr*cy;
+    R(0,2) = cr*sp*cy + sr*sy;
+
+    // Row 2
+    R(1,0) = sr*cp;
+    R(1,1) = sr*sp*sy + cr*cy;
+    R(1,2) = sr*sp*cy - cr*sy;
+
+    // Row 3
+    R(2,0) = -sp;
+    R(2,1) = cp*sy;
+    R(2,2) = cp*cy;
+
+    return R;
+    
+  }
+
+  Eigen::Vector3d Point2Point::getApproachVector(double roll, double pitch, double yaw) {
+    // Computes approach vector using rpy, as getApproachVector(i) does.
+    // This is unused and left only for reference. 
+    
+    Eigen::Vector3d ahat;
+    
+    double cr, sr; // cos/sin(roll)
+    double cp, sp; // cos/sin(roll)
+    double cy, sy; // cos/sin(roll)
+
+    // This is col(0) of rotation matrix as returned by Kinematics
+
+    cr = std::cos(roll);
+    sr = std::sin(roll);
+
+    cp = std::cos(pitch);
+    sp = std::sin(pitch);
+
+    cy = std::cos(yaw);
+    sy = std::sin(yaw);
+
+    ahat(0) = cr*cp;
+    ahat(1) = sr*cp;
+    ahat(2) = -sp;
+
+    return ahat;
+  }
+
+  Eigen::Vector3d Point2Point::getApproachVector(int i) {
+    return getEffRotation(i).col(0); // approach vector is x axis or exotica/Frame0Alwr_arm_6_link! 
+  }
+
+  Eigen::Matrix3d Point2Point::getEffRotation(int i) {
+    return Eigen::Map<Eigen::Matrix3d>(Kinematics[0].Phi(i).M.data).transpose(); // note, exotica/kdl returns transpose of matrix required
+  }
+
+  Eigen::Matrix3d Point2Point::getApproachVectorJacobian(double roll, double pitch, double yaw) {
+    // Approach vector jacobian (aJ)
+    //
+    // Let Phi = (r, p, y) be rpy euler angles
+    //     ahat = ahat(Phi) be approach vector of eff 
+    //
+    // Then aJ = d ahat(Phi) / d Phi 
+
+    Eigen::Matrix3d J;
+    
+    double cr, sr; // cos/sin(roll)
+    double cp, sp; // cos/sin(roll)
+    double cy, sy; // cos/sin(roll)
+
+    cr = std::cos(roll);
+    sr = std::sin(roll);
+
+    cp = std::cos(pitch);
+    sp = std::sin(pitch);
+
+    cy = std::cos(yaw);
+    sy = std::sin(yaw);
+
+    // Row 1
+    J(0,0) = -sr*cp;
+    J(0,1) = -cr*sp;
+
+    // Row 2
+    J(1,0) = cr*cp;
+    J(1,1) = -sr*sp;
+
+    // Row 3
+    J(2,1) = -cp;
+
+    return J;
+    
+  }
+
+  void Point2Point::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef J)
+  {
+
+    // Collect rpy angles
+    double roll, pitch, yaw;
+    Kinematics[0].Phi(0).M.GetRPY(roll, pitch, yaw);
+
+    // Compute eff postion and approach vector
+    Eigen::Vector3d e = getEffPosition(0);
+    Eigen::Vector3d ahat = getApproachVector(0);
+
+    // Setup
+    Eigen::Vector3d eap = e + ahat - point_;
+    Eigen::Vector3d L = point_ - e;
+    Eigen::Vector3d K = ahat.cross(eap);
+    Eigen::Matrix3d J_rpy = getApproachVectorJacobian(roll, pitch, yaw);
```

Comment:
This will surely suffer from Gimbal lock. Why do you need RPY? The alignment can be done with vector operations only.